File: fribidi-vs-unicode/bd16.py
######################################################################
# An implementation of bd16 in python
#
# Dov Grobgeld <dov.grobgeld@gmail.com>
# 2017-01-10 Tue
######################################################################

from collections import namedtuple

Element = namedtuple('Element',['char','pos'])
BidiProp = namedtuple('BidiProp',['type','opening'])

brack_prop = {
  '(': BidiProp(1,True),   ')' :BidiProp(1,False),
  '[': BidiProp(2,True),   ']' :BidiProp(2,False),
  '{': BidiProp(3,True),   '}' :BidiProp(3,False),
}
  
def find_pairings(text):
  stack = []  # Stack of elements
  pairs = []

  for p,c in enumerate(text):
    if c in brack_prop:
      prop = brack_prop[c]
      if prop.opening:
        stack += [Element(c,p)]
      else:
        sp = len(stack)-1
        for i in range(sp,-1,-1):
          se = stack[i]
          se_prop = brack_prop[se.char]
          if se_prop.type == prop.type:
            stack = stack[:i]
            pairs += [(se.pos, p)]
            break
  return tuple(sorted(pairs, key=lambda v: v[0]))

if __name__ == '__main__':
  Test = namedtuple('Test',['text','exp_ranges'])
  
  tests = [
    Test('a)b(c', tuple()),
    Test('a(b]c', tuple()),
    Test('a(b)c', ((1,3),)),
    Test('a(b[c)d]', ((1,5),)),
    Test('a(b]c)d', ((1,5),)),
    Test('a(b)c)d', ((1,3),)),
    Test('a(b(c)d', ((3,5),)),
    Test('a(b(c)d)',((1,7),(3,5))),
    Test('a(b{c}d)',((1,7),(3,5)))
  ]
  
  for t in tests:
    res = find_pairings(t.text)
    print res, t.exp_ranges == res
  

File: fribidi-vs-unicode/bd16.c
// A translation of the BD16.py
//
// Dov Grobgeld <dov.grobgeld@gmail.com>
// 2017-07-07 Fri

#include <stdlib.h>
#include <stdio.h>
#include <glib.h>
#include "bd16.h"

bool_t get_brack_prop(char c,
                      // output
                      brack_prop_t *prop)
{
  if (c=='(' || c==')')
    {
      prop->type = 1;
      prop->opening = (c=='(');
      return 1;
    }
  if (c=='[' || c==']')
    {
      prop->type = 2;
      prop->opening = (c=='[');
      return 1;
    }
  if (c=='{' || c=='}')
    {
      prop->type = 3;
      prop->opening = (c=='{');
      return 1;
    }
  return 0;
}

pairing_node_t *new_pairing_node(int start, int end)
{
  pairing_node_t *node =  g_new0(pairing_node_t, 1);
  node->start = start;
  node->end = end;
  node->next = NULL;

  return node;
}

element_t *new_element(char ch, int pos)
{
  element_t *node =  g_new0(element_t, 1);
  node->ch = ch;
  node->pos = pos;
  node->next = NULL;
  return node;
}

pairing_node_t *find_pairings(const char *text)
{
  pairing_node_t *pairs = NULL;
  element_t *pair_stack = NULL;

  const char *pch = text; // Assume zero terminated
  int pos = 0;
  while(*pch)
    {
      brack_prop_t brack_prop;
      if (get_brack_prop(*pch,
                         // output
                         &brack_prop))
        {
          if (brack_prop.opening)
            {
              // push onto the pair stack
              element_t *node = new_element(*pch, pos);
              node->next = pair_stack;
              pair_stack = node;
            }
          else
            {
              element_t *pst = pair_stack;
              brack_prop_t se_brack_prop;
              while(pst)
                {
                  get_brack_prop(pst->ch,
                                 // output
                                 &se_brack_prop);
                  if (se_brack_prop.type == brack_prop.type)
                    {
                      // pop until we hit pst
                      while(pair_stack != pst)
                        {
                          element_t *p = pair_stack;
                          pair_stack = pair_stack->next;
                          free(p);
                        }

                      pairing_node_t *pair = new_pairing_node(pst->pos, pos);
                      pair->next = pairs;
                      pairs = pair;
                      pair_stack = pair_stack->next;
                      free(pst);
                      break;
                    }
                  pst=pst->next;
                }
            }
        }
      pch++;
      pos++;
    }

  // No pair found. just clean the stack
  while (pair_stack)
    {
      element_t *p = pair_stack;
      pair_stack = pair_stack->next;
      free(p);
    }
  
  // Do we need to sort?
  return pairs;
}

